micro-rdk/src/common/data_manager.rs

```diff
@@ -269,13 +288,70 @@ pub enum DataSyncError {
     DataStoreError(#[from] DataStoreError),
     #[error(transparent)]
     MessageDecodingError(#[from] prost::DecodeError),
+    #[error("time correction resulted in out of bounds duration")]
+    TimeOutOfBoundsError,
+    #[error("current time unset")]
+    NoCurrentTime,
 }
 
 pub struct DataSyncTask<StoreType> {
     store: Rc<AsyncMutex<StoreType>>,
     resource_method_keys: Vec<ResourceMethodKey>,
     sync_interval: Duration,
     part_id: String,
+    // used for time correcting stored data before upload, see DataSyncTask::run
+    // and create_time_corrected_reading below
+    robot_start_time: Instant,
+}
+
+fn get_time_to_subtract(
+    robot_start_time: Instant,
+    stored_time: Timestamp,
+) -> Result<chrono::Duration, DataSyncError> {
+    let stored_time_dur = Duration::new(stored_time.seconds as u64, stored_time.nanos as u32);
+    let time_to_subtract = robot_start_time.elapsed() - stored_time_dur;
+    let time_to_subtract = chrono::Duration::new(
+        time_to_subtract.as_secs() as i64,
+        time_to_subtract.subsec_nanos(),
+    )
+    .ok_or(DataSyncError::TimeOutOfBoundsError)?;
+    Ok(time_to_subtract)
+}
+
+fn create_time_corrected_reading(
```

Comment:
This function doesn't create a reading though: it creates a thing which adjusts readings. `create_reading_time_corrector`?